File: core/prime.ts
import { readStore, applyDecay, filterHints } from "../store/reader.ts"
import type { Hint } from "../store/types.ts"

export interface PrimeResult {
  domain: string
  hintsLoaded: number
  promptBlock: string
}

export function prime(domain: string): PrimeResult {
  const raw = readStore(domain)
  const decayed = applyDecay(raw)
  const hints = filterHints(decayed)

  if (hints.length === 0) {
    return { domain, hintsLoaded: 0, promptBlock: "" }
  }

  const lines = hints.map((h) => formatHint(h))
  const promptBlock = [
    `KNOWN HINTS FOR ${domain}:`,
    ...lines,
    "",
  ].join("\n")

  return { domain, hintsLoaded: hints.length, promptBlock }
}

function formatHint(h: Hint): string {
  const conf = Math.round(h.confidence * 100)
  return `- [${conf}% confident] ${h.note}. ${h.action}`
}

export function buildGoal(originalGoal: string, primeResult: PrimeResult): string {
  if (!primeResult.promptBlock) return originalGoal
  return `${primeResult.promptBlock}\nTASK: ${originalGoal}`
}

File: mycelium-py/src/mycelium/prime.py
from __future__ import annotations

from .store import apply_decay, filter_hints, read_store
from .types import Hint, PrimeResult

_STOP_RULES = """RULES:
- Follow the hints above — do not rediscover what is already known
- If a hint redirects you to a different site or method, go there directly without attempting the original first
- If you hit a login wall, cookie banner, or access block that has no hint: stop and return what you have, do not retry"""


def _format_hint(h: Hint) -> str:
    conf = round(h.confidence * 100)
    prefix = f"- [SHORTCUT, {conf}% confident]" if h.type == "flow" else f"- [{conf}% confident]"
    return f"{prefix} {h.note} → {h.action}"


def prime(domain: str, goal: str | None = None) -> PrimeResult:
    store = apply_decay(read_store(domain))
    hints = filter_hints(store, goal)

    if not hints:
        return PrimeResult(domain=domain, hints_loaded=0, prompt_block="")

    lines = [_format_hint(h) for h in hints]
    prompt_block = "\n".join(
        [
            f"IMPORTANT — KNOWN HINTS FOR {domain} (follow these, do not rediscover):",
            *lines,
            "",
        ]
    )
    return PrimeResult(domain=domain, hints_loaded=len(hints), prompt_block=prompt_block)


def build_goal(original_goal: str, primed: PrimeResult) -> str:
    if not primed.prompt_block:
        return original_goal
    return f"{primed.prompt_block}\n{_STOP_RULES}\n\nTASK: {original_goal}"
